numpy/core/_internal.pyi

```diff
@@ -1,18 +1,37 @@
-from typing import Any
+from typing import Any, TypeVar, Type, overload, Optional, Generic
+import ctypes as ct
 
-# TODO: add better annotations when ctypes is stubbed out
+from numpy import ndarray
 
-class _ctypes:
+_CastT = TypeVar("_CastT", bound=ct._CanCastTo)  # Copied from `ctypes.cast`
+_CT = TypeVar("_CT", bound=ct._CData)
+_PT = TypeVar("_PT", bound=Optional[int])
+
+# TODO: Let the likes of `shape_as` and `strides_as` return `None`
+# for 0D arrays once we've got shape-support
+
+class _ctypes(Generic[_PT]):
+    @overload
+    def __new__(cls, array: ndarray[Any, Any], ptr: None = ...) -> _ctypes[None]: ...
+    @overload
+    def __new__(cls, array: ndarray[Any, Any], ptr: _PT) -> _ctypes[_PT]: ...
+
+    # NOTE: In practice `shape` and `strides` return one of the concrete
+    # platform dependant c_int64-based array-types (`c_int`, `c_long` or
+    # `c_longlong`)
     @property
-    def data(self) -> int: ...
+    def data(self) -> _PT: ...
     @property
-    def shape(self) -> Any: ...
+    def shape(self) -> ct.Array[ct.c_int64]: ...
     @property
-    def strides(self) -> Any: ...
-    def data_as(self, obj: Any) -> Any: ...
-    def shape_as(self, obj: Any) -> Any: ...
-    def strides_as(self, obj: Any) -> Any: ...
-    def get_data(self) -> int: ...
-    def get_shape(self) -> Any: ...
-    def get_strides(self) -> Any: ...
-    def get_as_parameter(self) -> Any: ...
+    def strides(self) -> ct.Array[ct.c_int64]: ...
+    @property
+    def _as_parameter_(self) -> ct.c_void_p: ...
+
+    def data_as(self, obj: Type[_CastT]) -> _CastT: ...
+    def shape_as(self, obj: Type[_CT]) -> ct.Array[_CT]: ...
+    def strides_as(self, obj: Type[_CT]) -> ct.Array[_CT]: ...
+    def get_data(self) -> _PT: ...
+    def get_shape(self) -> ct.Array[ct.c_int64]: ...
+    def get_strides(self) -> ct.Array[ct.c_int64]: ...
+    def get_as_parameter(self) -> ct.c_void_p: ...
```

Comment:
It's tempting to deliberately exclude these, since they're not really intended to be public API.

Reply:
Looking at source code, it seems that methods are there simply to preserve backwards compatibility.
If so, would it be worthwhile to simply deprecate them?
https://github.com/numpy/numpy/blob/bfb1ac86588456a78af63fa314b768f4ebc764ff/numpy/core/_internal.py#L353-L357

Reply:
(after pinging the mailing list, that is)

Reply:
I'd be in favor. I added that compatibility comment in 79baef7a80769842f1c7f14e31056e4750a122e8. These functions date back to at least eee00f8f7e15592a048c8b841aef9ea81faa0fda, and in all cases they've just been an implementation detail for the `property`s that follow; I doubt anyone on the mailing list will even have heard of them (they don't appear in the docs).

Reply:
Method annotations removed as of https://github.com/numpy/numpy/pull/19029/commits/44c05b32d119353eb8ba28edfb042602b591e590.